Notebooks/SentinelUtilities/SentinelAnomalyLookup/anomaly_lookup.py

```diff
@@ -0,0 +1,241 @@
+from System import *
+from Microsoft.Azure.CIS.Notebooks.AnomalyLookup import *
+from Microsoft.Azure.CIS.Notebooks.LogHelper import *
+from .anomaly_lookup_view_helper import *
+
+
+class AnomalyLookup(object):
+    def __init__(self, workspace_id, la_data_client):
+        self.workspace_id = workspace_id
+        self.la_data_client = la_data_client
+        self.logger = AILogger();
+
+    def query_table_list(self):
+        return self.query_loganalytics(KqlLibrary.ListTables())
+
+    def query_loganalytics(self, query):
+        res = self.la_data_client.query(self.workspace_id, QueryBody(query=query))
+        json = res.as_dict()
+        cols = json_normalize(json['tables'][0], 'columns')
+        df = json_normalize(json['tables'][0], 'rows')
+        if (df.shape[0] != 0):
+            df.columns = cols.name
+        return df
+
+    def construct_related_queries(self, dfAnomalies):
+        if (dfAnomalies.shape[0] == 0):
+            return
+
+        queries = ''
+        for tbl in dfAnomalies.Table.unique():
+
+            curTableAnomalies = dfAnomalies.ix[dfAnomalies.Table == tbl,:]
+            query = """{tbl} | where TimeGenerated > ago(30d) | where ingestion_time() > datetime({maxTimestamp})-1d and ingestion_time() < datetime({maxTimestamp}) | where {entCol} has "{qEntity}" | where """.format(**{
+                'tbl': tbl,
+                'qTimestamp': curTableAnomalies.qTimestamp.iloc[0],
+                'maxTimestamp': curTableAnomalies.maxTimestamp.iloc[0],
+                'entCol': curTableAnomalies.entCol.iloc[0],
+                'qEntity': curTableAnomalies.qEntity.iloc[0]
+            })
+
+            for j, row in curTableAnomalies.iterrows():
+                query += " {col} == to{colType}(\"{colVal}\") or".format(**{
+                    'col': row.colName,
+                    'colType': (row.colType) if 'colType' in row.keys() else 'string',
+                    'colVal': row.colVal
+                })
+
+            query = query[:-2] # drop the last or
+            query += " | take 1000" # limit the output size
+
+            query = query.replace("\\","\\\\")
+            #print(query + "\n\n")
+
+            queries += query
+        return queries
+
+    def get_timewindow(self, qEntity, qTimestamp, entCol, tbl):
+        # find the relevant time window for analysis
+        winStart = 0
+        minTimestamp = None
+        delta = None
+        maxTimestamp = None
+        longMinTimestamp = None
+        for f in range(-30, 0, 1):
+            dfTimeRange = self.query_loganalytics(KqlLibrary.TimeWindowQuery(tbl, qTimestamp, entCol, qEntity, f, f+1, 'd'))
+        
+            if (dfTimeRange.shape[0] > 0):
+                winStart = f
+                break
+
+        dtQTimestamp = pd.to_datetime(qTimestamp)
+        ind2now = dt.datetime.utcnow() - dtQTimestamp
+        if (winStart < -3):
+            if (ind2now > dt.timedelta(days=1)):
+                delta = '1d'
```

Comment:
move to outer "if" ?

Reply:
Code from Researcher, will let it as it is.